scrapy_po/utils.py

```diff
@@ -1,7 +1,122 @@
 # -*- coding: utf-8 -*-
+import inspect
+from typing import Tuple, Dict, Type, Callable
+
+import andi
+from scrapy.http import Response
+from scrapy.utils.defer import maybeDeferred_coro
+from twisted.internet.defer import inlineCallbacks, returnValue
+
+from scrapy_po.webpage import Injectable
+from scrapy_po.page_input_providers import providers
+
 
 def get_callback(request, spider):
-    """ Get request.callback of a scrapy.Request, as a callable """
+    """Get request.callback of a scrapy.Request, as a callable."""
     if request.callback is None:
         return getattr(spider, 'parse')
     return request.callback
+
+
+class DummyResponse(Response):
+
+    def __init__(self, url, request=None):
+        super(DummyResponse, self).__init__(url=url, request=request)
+
+
+def is_callback_using_response(callback: Callable):
+    """Check whether the request's callback method is going to use response."""
+    spec = inspect.getfullargspec(callback)
+    try:
+        arg_name = spec.args[1]  # first index is self, second is response
+    except IndexError:
+        # Parse method is probably using *args and **kwargs annotation.
+        # Let's assume response is going to be used.
+        return True
+
+    if arg_name not in spec.annotations:
+        # There's no type annotation, so we're probably using response here.
+        return True
+
+    if issubclass(spec.annotations[arg_name], DummyResponse):
+        # Type annotation is DummyResponse, so we're probably NOT using it.
+        return False
+
+    # Type annotation is not DummyResponse, so we're probably using it.
+    return True
+
+
+def are_dependencies_using_response(plan: andi.Plan):
```

Comment:
a minor thing, but what do you think about making `providers` an argument for this function, to get further from singletons, and potentially simplify testing?

Reply:
Refactored.